File: src/main/java/com/rs/java/game/player/actions/skills/crafting/glass/GlassBlowingData.java
package com.rs.java.game.player.actions.skills.crafting.glass;

import com.rs.kotlin.Rscm;
import com.rs.java.game.player.actions.skills.crafting.leather.ReqItem;

public enum GlassBlowingData {

    MOLTEN_GLASS(
            "item.molten_glass",

            product("item.beer_glass", 1, 17.5),
            product("item.candle_lantern", 4, 19),
            product("item.oil_lamp", 12, 25),
            product("item.vial", 33, 35),
            product("item.fishbowl", 42, 42.5),
            product("item.unpowered_orb", 46, 52.5),
            product("item.lantern_lens", 49, 55),
            product("item.empty_light_orb", 87, 70)
    );

    private final Object base;
    private final GlassProduct[] products;

    GlassBlowingData(Object base, GlassProduct... products) {
        this.base = base;
        this.products = products;
    }

    private static GlassProduct product(Object id, int lvl, double xp, ReqItem... req) {
        return new GlassProduct(id, lvl, xp, req);
    }

    public int getBase() {
        if (base instanceof Integer)
            return (Integer) base;
        return Rscm.lookup((String) base);
    }

    public static GlassBlowingData getGlassData(int id1, int id2) {

        int pipe = 1785; // glassblowing pipe
        int molten = Rscm.lookup("item.molten_glass");

        boolean hasPipe = id1 == pipe || id2 == pipe;
        boolean hasMolten = id1 == molten || id2 == molten;

        if (hasPipe && hasMolten)
            return MOLTEN_GLASS;

        return null;
    }


    public GlassProduct[] getProducts() {
        return products;
    }
}

File: src/main/java/com/rs/java/game/player/actions/skills/crafting/glass/GlassBlowingData.kt
package com.rs.java.game.player.actions.skills.crafting.glass

import com.rs.kotlin.rscm.RscmResolver

enum class GlassBlowingData(
    val baseRef: Any,
    val products: Array<GlassProduct>,
) {
    MOLTEN_GLASS(
        "item.molten_glass",
        arrayOf(
            GlassProduct("item.beer_glass", 1, 17.5),
            GlassProduct("item.candle_lantern", 4, 19.0),
            GlassProduct("item.oil_lamp", 12, 25.0),
            GlassProduct("item.vial", 33, 35.0),
            GlassProduct("item.fishbowl", 42, 42.5),
            GlassProduct("item.unpowered_orb", 46, 52.5),
            GlassProduct("item.lantern_lens", 49, 55.0),
            GlassProduct("item.empty_light_orb", 87, 70.0),
        ),
    ),
    ;

    companion object {
        private const val GLASSBLOWING_PIPE = 1785

        private val baseLookup =
            RscmResolver.buildObjectIdMap(entries) { listOf(it.baseRef) }

        private val productLookup =
            RscmResolver.buildObjectIdMap(
                entries.flatMap { it.products.toList() },
            ) { listOf(it.idRef) }

        fun resolveBase(id: Int): GlassBlowingData? = baseLookup[id]

        fun resolveProduct(id: Int): GlassProduct? = productLookup[id]

        fun getGlassData(
            id1: Int,
            id2: Int,
        ): GlassBlowingData? {
            val data = resolveBase(id1) ?: resolveBase(id2) ?: return null

            val hasPipe = id1 == GLASSBLOWING_PIPE || id2 == GLASSBLOWING_PIPE

            return if (hasPipe) data else null
        }
    }
}
